### ISSUE TITLE
More volatile casts might be needed for OCaml 5.0

### ISSUE THREAD

**Issue by gadmm** [ocaml/ocaml#11426]
One should audit the new runtime to see if any other places in the new runtime need volatile casts (`Field` was only meant as an example of problem). According to @kayceesrk this has not been checked yet.

- [x] Are other macros similar to `Field` in their treatment?
- [x] I still think something like `typedef volatile value * value_ref` is useful to let users not think about the `volatile` implementation detail (which so far is not in the memory model). This might depend on usages of `value *` in the wild.
- [ ] In fact, even OCaml uses (rare) value pointers in its API. See `caml_named_value`, which incidentally should certainly return `const volatile value *` instead of `const value *` given that parallel mutations are possible via `caml_register_named_value`. (Though it would be better, if possible, to avoid this data race while keeping the same API in the first place).
- [ ] Audit the new runtime to see if any other places in the new runtime need volatile casts.

_Originally posted by @gadmm in https://github.com/ocaml/ocaml/issues/11255#issuecomment-1148781023_

**Comment by edwintorok** [comment]
At a quick glance `memory.h` seems to need more `volatile`, especially the global root registration mechanism, which might imply that these values get modified outside of the caller's control.
```
memory.h:CAMLextern void caml_modify (volatile value *, value);
memory.h:CAMLextern void caml_initialize (volatile value *, value);
memory.h:  value *tables [5];
memory.h:CAMLextern void caml_register_global_root (value *);
memory.h:CAMLextern void caml_remove_global_root (value *);
memory.h:   [caml_modify_generational_global_root].  The [value *] pointer
memory.h:CAMLextern void caml_register_generational_global_root (value *);
memory.h:CAMLextern void caml_remove_generational_global_root (value *);
memory.h:CAMLextern void caml_modify_generational_global_root(value *r, value newval);
```

The 'tables' in the roots block may not necessarily need a volatile: the GC must already be able to deal with OCaml 'value' being spilled onto the stack or stored in a register by the compiler when in C bindings, e.g. the value may have moved after a 'caml_leave_blocking_section', and also that pointer is reachable from a global pointer, so the compiler must already be aware that it can change outside its control ... unless all the casts break GCC's type punning assumptions.

However it might seem simpler to modify `typedef intnat value` to include a `volatile` (perhaps only in C bindings and not when inside the runtime itself where more careful could be made) rather than chasing down which pointer needs a volatile and which one doesn't. If there aren't strong objections against a change like that I could set up a private branch to assess the impact of such a change on C bindings, i.e. to see whether it would cause more compiler warnings or errors, and whether it would it change the generated assembly in a meaningful way.

**Comment by OlivierNicole** [comment]
I can work on this. I think studying the empirical data given by running the test suite with ThreadSanitizer can be useful in signaling places where volatile casts might be missing. Could I be assigned to this issue?

**Comment by gadmm** [comment]
@edwintorok Sorry for the late reply—I think those occurrences of `value *` in global roots functions are fine as is. The volatile is only needed for values that reside on the OCaml heap. Races involving the location recorded in the root can be dealt with by ensuring ownership/synchronization of the root on the C side. (We had the exact same question in the RFC for boxroot.)

@OlivierNicole It is great if ThreadSanitizer can help. But as far as I understand ThreadSanitizer cannot guarantee the absence of bugs, hence why a systematic review of the runtime will still be needed.

**Comment by OlivierNicole** [comment]
> @OlivierNicole It is great if ThreadSanitizer can help. But as far as I understand ThreadSanitizer cannot guarantee the absence of bugs, hence why a systematic review of the runtime will still be needed.

Definitely. It can only give hints, or complement the audits made by fallible humans.

**Comment by OlivierNicole** [comment]
I started auditing the problematic macros other than `Field`. By problematic macros, I mean macros that:

- Are a documented part of the FFI, or are *de facto* used in many OCaml projects
- Can be used to perform plain reads on values that may be mutated in parallel **by OCaml code or C code from the runtime**.

Plain reads that may only race with other C code are the responsibility of the person writing the C stubs.

It turns out that looking only at the macros documented in [Interfacing C with OCaml](https://ocaml.org/manual/intfc.html), many are problematic:

- `Tag_val`: returns the tag of a value. May race with the runtime updating the header for GC or lazy evaluation.
- ~~`Byte`: returns the n-th byte of the value (not counting the header).~~
- ~~`Byte_u`: same, but returns an unsigned char~~ Edit: these two are pretty much a lost cause given how they are used throughout existing code. I would say that when someone does byte-per-byte operations on a value, they should know what they are doing and we won’t give any guarantees when the value is mutated concurrently.
- `String_val` (if dereferenced): returns a `const char *` from a `value`. Strings are in principle immutable in OCaml, so to get races with this one, one would have to use it on a value that is not a string. I think we should probably not support such mad uses.
- `Bytes_val` (if dereferenced): returns an `unsigned char *` from a `value`.
- ~~`Double_val`: unboxes a boxed float. Rather pervasive in the runtime and probably in user code too.~~ Scratch that, floats are actually immutable if you set aside the case of flat float arrays.
- `Double_field` and `Double_array_field`: reads a field in a float-only array or record.

**Suggestion:** we should add a `volatile` cast to the macros `Tag_val`, `Byte`, `Byte_u`, `Bytes_val`, `Double_val` and `Double_field`.

**Threats:** as for `Field`, taking the address of such constructs, as in `&Byte(v, n)`, will have a `volatile`-qualified type incompatible with its previous type, which can cause errors and thus not be fully compatible. The occurrence of such incompatibilities in real code should be studied.

Auditing work left to do (in progress): audit undocumented macros that may be used by some developers; evaluate usage of value pointers in the OCaml API, as described in the original message of this issue.

**Comment by OlivierNicole** [comment]
Status update: I have added those missing `volatile`s on a branch at https://github.com/OlivierNicole/ocaml/tree/more_volatile. I am waiting for cluster availability to do an opam-health-check of this.

I also added `volatile`s to a number of _undocumented_ macros of `caml/mlvalues.h` (some of which are used in some opam packages). Those problematic macros are:

- `Hd_hp`
- `Hp_val` if dereferenced
- `Hp_op` if dereferenced
- `Hp_bp` if dereferenced
- `Op_hp` if dereferenced
- `Bp_hp` if dereferenced

- `Wosize_hp`
- `Whsize_hp`
- `Whsize_bp`
- `Whsize_hp`
- `Whsize_val`
- `Whsize_bp`
- `Bhsize_hp`

Remark: the macros that read a size can be fixed without backward-compatibility problems.

- `Tag_hp`
- `Op_val`
- `Bp_val` if dereferenced

**Comment by OlivierNicole** [comment]
@gadmm I believe the first two tasks of this issue can be marked as completed?

**Comment by gadmm** [comment]
To be honest I lost track of the work that has been done. I can tell that you did a good amount of work going in the right direction. But I would not know how to become confident that the audit is complete without doing a deep dive by myself, and I am not available for this. Could someone do a meta-review maybe?

**Comment by OlivierNicole** [comment]
For the record,

- Task 1 has been addressed, I believe, by #12512 
- Regarding task 2, the type `value_ptr` has been added in #12030 
- Regarding task 3, no systematic audit has been made yet of API functions dealing with `value *` pointers as far as I know.

**Comment by OlivierNicole** [comment]
I would like to reiterate my proposal of marking the first two tasks of this issue as completed, and ask for a third opinion if @gadmm disagrees. It’s harder to keep track of progress if we don’t update our logs.

Task 2 is clearly done, task 1 has been explicitly addressed by a PR that was reviewed and merged. The request of a meta-review (reviewing that something that was already reviewed does address what it was supposed to address) strikes me as a request of unnecessary additional work.

**Comment by gadmm** [comment]
My message was not an opposition. I have ticked the first two boxes (and I believe anyone with a green mark can do so). I trust you that in the case of macros the situation has been reviewed (in the case of mlvalues.h? or all header files?). I added a fourth box so that it is clearer that the scope of this issue is more than the items in boxes.

I was rather stating that I lost track and do not have much time to get back to these issues. And there is indeed some nontrivial work to review whether the whole runtime has been audited for these sort of issues. The problem is the lack of active contributors to OCaml willing/able to contribute to the review of the new multicore runtime which is still going on.

By the way, the commit cross-referenced just above might indicate that there are backwards-compatibility issues with the volatile changes. (If this is the case @tbrk, the ocaml devs will probably be interested in feedback, if you like.)
